oval/redhat/redhat.go

```diff
@@ -0,0 +1,110 @@
+package redhat
+
+import (
+	"bytes"
+	"compress/bzip2"
+	"encoding/xml"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/afero"
+	"golang.org/x/xerrors"
+	"gopkg.in/cheggaaa/pb.v1"
+
+	"github.com/aquasecurity/vuln-list-update/utils"
+)
+
+const (
+	ovalDir   = "oval"
+	redhatDir = "redhat"
+	urlFormat = "https://www.redhat.com/security/data/oval/v2/RHEL%s/rhel-%s.oval.xml.bz2"
+	retry     = 5
+)
+
+var (
+	releases = []string{"6", "7", "8"}
+)
+
+type Config struct {
+	VulnListDir string
+	URLFormat   string
+	AppFs       afero.Fs
+	Retry       int
+}
+
+func NewConfig() Config {
+	return Config{
+		VulnListDir: utils.VulnListDir(),
+		URLFormat:   urlFormat,
+		AppFs:       afero.NewOsFs(),
+		Retry:       retry,
+	}
+}
+
+func (c Config) Update() error {
+	log.Println("Fetching Red Hat OVAL data...")
+	for _, release := range releases {
+		if err := c.update(release); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c Config) update(release string) error {
+	url := fmt.Sprintf(c.URLFormat, release, release)
+	res, err := utils.FetchURL(url, "", c.Retry)
+	if err != nil {
+		return xerrors.Errorf("failed to fetch Red Hat OVAL: %w", err)
+	}
+
+	bzr := bzip2.NewReader(bytes.NewBuffer(res))
+
+	log.Printf("Updating Red Hat %s OVAL data...\n", release)
+	ovalroot := Root{}
+	if err = xml.NewDecoder(bzr).Decode(&ovalroot); err != nil {
+		return xerrors.Errorf("failed to unmarshal Red Hat OVAL XML: %w", err)
+	}
+
+	dir := filepath.Join(ovalDir, redhatDir, release)
+	bar := pb.StartNew(len(ovalroot.Definitions.Definitions))
+	for _, def := range ovalroot.Definitions.Definitions {
+		titles := strings.Fields(def.Title)
+		title := strings.Trim(titles[0], ":")
+		if err = c.saveRHSAPerYear(dir, title, def); err != nil {
+			return xerrors.Errorf("unable to save RHEL advisory: %w", err)
+		}
+		bar.Increment()
+	}
+	bar.Finish()
+	return nil
+}
+
+func (c Config) saveRHSAPerYear(dirName string, rhsaID string, data interface{}) error {
+	// e.g. RHSA-2018:0094
+	s := strings.Split(rhsaID, ":")
+	if len(s) != 2 {
+		log.Printf("invalid RHSA-ID format: %s\n", rhsaID)
+		return nil
```

Comment:
Why do we return `nil` back if the format was invalid?

Reply:
I've added a failing test to showcase the behaviour seen: https://github.com/aquasecurity/vuln-list-update/pull/19/commits/39257bba851c947404937720b046128ba37bfed6

Reply:
`afero` is used for FileSystem mocking. If you use NewMemMapFs, files are saved on a memory, not on a local file system. So, this is unit testing.
https://github.com/spf13/afero#step-2-declare-a-backend

Reply:
Red Hat data is sometimes broken. In that case, if we return err, the entire update will be stopped. For example, the Red Hat xml has 1,000 vulnerabilities. Even if 1 vulnerability is broken, 999 vulnerabilities should be saved. If we return err, the update of 1,000 vulnerabilities will fail.

Reply:
Hmm I see. I think in that case we should at least distinguish between a good return `return nil` vs a bad return `return something else` in order clarify to the caller what happened. As it stands there's no way to test if the call to `saveRHSAPerYear` was a success or a failure because return value is the same for both.

Reply:
I've added a commit to help that: https://github.com/aquasecurity/vuln-list-update/pull/19/commits/b8e8adf4966f1c05a2f839f9fefc88f3c81b1b62